Library/Homebrew/dev-cmd/livecheck.rb

```diff
@@ -0,0 +1,459 @@
+  private_constant :WATCHLIST_PATH
+
+  GITHUB_SPECIAL_CASES = %w[
+    api.github.com
+    /latest
+    mednafen
+    camlp5
+    kotlin
+    osrm-backend
+    prometheus
+    pyenv-virtualenv
+    sysdig
+    shairport-sync
+    yuicompressor
+  ].freeze
+  private_constant :GITHUB_SPECIAL_CASES
+
+  UNSTABLE_VERSION_KEYWORDS = %w[
+    alpha
+    beta
+    bpo
+    dev
+    experimental
+    prerelease
+    preview
+    rc
+  ].freeze
+  private_constant :UNSTABLE_VERSION_KEYWORDS
+
+  NO_VERSIONS_MSG = "Unable to get versions"
+  HEAD_ONLY_MSG = "HEAD only formula must be installed to be livecheckable"
+  private_constant :NO_VERSIONS_MSG, :HEAD_ONLY_MSG
+
+  @livecheck_strategy_names = {}
+
+  def livecheck_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `livecheck` [<formulae>]
+
+        Check for newer versions of formulae from upstream.
+
+        If no formula argument is passed, the list of formulae to check is taken from `HOMEBREW_LIVECHECK_WATCHLIST`
+        or `~/.brew_livecheck_watchlist`.
+      EOS
+      switch :verbose
+      switch :quiet
+      switch :debug
+      switch "--full-name",
+             description: "Print formulae with fully-qualified names."
+      flag   "--tap=",
+             description: "Check the formulae within the given tap, specified as <user>`/`<repo>."
+      switch "--installed",
+             description: "Check formulae that are currently installed."
+      switch "--json",
+             description: "Output informations in JSON format."
+      switch "--all",
+             description: "Check all available formulae."
+      switch "--newer-only",
+             description: "Show the latest version only if it's newer than the formula."
+      conflicts "--debug", "--json"
+      conflicts "--tap=", "--all", "--installed"
+    end
+  end
+
+  def livecheck
+    args = livecheck_args.parse
+
+    if args.debug? && args.verbose?
+      puts ARGV.inspect
+      puts args
+      puts ENV["HOMEBREW_LIVECHECK_WATCHLIST"] if ENV["HOMEBREW_LIVECHECK_WATCHLIST"].present?
+    end
+
+    formulae_to_check =
+      if args.tap
+        Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
+      elsif args.installed?
+        Formula.installed
+      elsif args.all?
+        Formula.full_names.map { |name| Formula[name] }
+      elsif !args.formulae.empty?
+        args.formulae
+      elsif File.exist?(WATCHLIST_PATH)
+        begin
+          File.readlines(WATCHLIST_PATH).each_with_object([]) do |word, memo|
+            memo << Formulary.factory(word.chomp) unless word.start_with?("#")
+          end
+        rescue Errno::ENOENT => e
+          onoe e
+        end
+      end
+
+    if formulae_to_check.nil? || (formulae_to_check.is_a?(Array) && formulae_to_check.empty?)
+      onoe "No formulae to check."
+      return
+    end
+
+    # Identify any non-homebrew/core taps in use for current formulae
```

Comment:
I'd say everything here and below should be in the `Homebrew::Livecheck` module.

Reply:
With this in mind, we should probably move the constants that come before `def livecheck_args` (e.g., `WATCHLIST_PATH`, `GITHUB_SPECIAL_CASES`, etc.) into the `Homebrew::Livecheck` module as well, correct? If so, `@livecheck_strategy_names` could also be in `Homebrew::Livecheck`.